File: week1_baseline/ruby/12_context/lib/boukensha/backends/anthropic.rb
require_relative "base"

# https://platform.claude.com/docs/en/api/beta/messages/create
module Boukensha
  module Backends
    class Anthropic < Base
      BASE_URL = "https://api.anthropic.com/v1/messages"
      MODELS = {
        # thinking, but not adapative must set budget_tokens
        # does not support effort flag
        "claude-haiku-4-5" => {
          context_window: 200_000,
          cost_per_million: { input: 1.0, output: 5.0 },
          usage_unit: :tokens
        },
        # supports adapative thinking and effort flag
        # non-adpative thinking is deprecated for sonnet 
        "claude-sonnet-4-6" => {
          context_window: 1_000_000,
          cost_per_million: { input: 3.0, output: 15.0 },
          usage_unit: :tokens
        },
        # supports adapative thinking and effort flag
        # non-adaptive thinking is deprecated for opus
        "claude-opus-4-8" => {
          context_window: 1_000_000,
          cost_per_million: { input: 5.0, output: 25.0 },
          usage_unit: :tokens
        }
      }.freeze

      def initialize(api_key:, model:)
        @api_key = api_key
        configure_model(model)
      end

      def to_messages(messages)
        messages.map do |msg|
          case msg.role
          when :tool_result
            {
              role: "user",
              content: [{
                type: "tool_result",
                tool_use_id: msg.tool_use_id,
                content: msg.content
              }]
            }
          when :assistant
            { role: "assistant", content: assistant_content(msg.content) }
          else
            { role: msg.role.to_s, content: msg.content }
          end
        end
      end

      def to_tools(tools)
        tools.values.map do |tool|
          {
            name: tool.name,
            description: tool.description,
            input_schema: {
              type: "object",
              properties: tool.parameters,
              required: tool.parameters.keys.map(&:to_s)
            }
          }
        end
      end

      def to_payload(context, max_output_tokens: 1024, tools: nil)
        {
          model: @model,
          system: context.system,
          max_tokens: max_output_tokens,
          tools: tools.nil? ? to_tools(context.tools) : tools,
          messages: to_messages(context.messages)
        }
      end

      def headers
        {
          "Content-Type"      => "application/json",
          "x-api-key"         => @api_key,
          "anthropic-version" => "2023-06-01"
        }
      end

      def url
        BASE_URL
      end

      # Normalizes an Anthropic Messages API response into the common shape
      # (see Backends::Base for the full content-block contract). Anthropic's
      # native thinking/redacted_thinking blocks are mapped to "reasoning"
      # blocks, preserving the signature so they can be echoed back unchanged
      # (the API rejects modified thinking blocks when continuing on the same
      # model).
      def parse_response(response)
        stop_reason = response["stop_reason"] == "tool_use" ? "tool_use" : "end_turn"
        content     = (response["content"] || []).map { |block| normalize_block(block) }
        { stop_reason: stop_reason, content: content }
      end

      private

      def normalize_block(block)
        case block["type"]
        when "thinking"
          { "type" => "reasoning", "text" => block["thinking"].to_s, "signature" => block["signature"] }
        when "redacted_thinking"
          { "type" => "reasoning", "text" => "", "redacted" => true, "signature" => block["data"] }
        else
          block
        end
      end

      # Rebuilds Anthropic assistant content from normalized blocks (the inverse
      # of parse_response). Text-only turns are stored as a bare String and pass
      # through unchanged; "reasoning" blocks are re-emitted as native
      # thinking/redacted_thinking blocks so signatures round-trip intact.
      def assistant_content(content)
        return content if content.is_a?(String)

        content.map { |block| denormalize_block(block) }
      end

      def denormalize_block(block)
        return block unless block["type"] == "reasoning"

        if block["redacted"]
          { "type" => "redacted_thinking", "data" => block["signature"] }
        else
          { "type" => "thinking", "thinking" => block["text"].to_s, "signature" => block["signature"] }
        end
      end
    end
  end
end

File: week1_baseline/python/12_context/boukensha/backends/anthropic.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, ClassVar

from .base import Base

if TYPE_CHECKING:
    from ..context import Context
    from ..message import Message
    from ..tool import Tool


class Anthropic(Base):
    BASE_URL = "https://api.anthropic.com/v1/messages"
    MODELS: ClassVar[dict[str, dict[str, Any]]] = {
        "claude-haiku-4-5": {
            "context_window": 200_000,
            "cost_per_million": {"input": 1.0, "output": 5.0},
            "usage_unit": "tokens",
        },
        "claude-sonnet-4-6": {
            "context_window": 1_000_000,
            "cost_per_million": {"input": 3.0, "output": 15.0},
            "usage_unit": "tokens",
        },
        "claude-opus-4-8": {
            "context_window": 1_000_000,
            "cost_per_million": {"input": 5.0, "output": 25.0},
            "usage_unit": "tokens",
        },
    }

    def __init__(self, *, api_key: str, model: str) -> None:
        self.api_key = api_key
        self._configure_model(model)

    def to_messages(self, messages: list[Message]) -> list[dict[str, Any]]:
        result = []
        for msg in messages:
            if msg.role == "tool_result":
                result.append(
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "tool_result",
                                "tool_use_id": msg.tool_use_id,
                                "content": msg.content,
                            }
                        ],
                    }
                )
            elif msg.role == "assistant":
                result.append({"role": "assistant", "content": self._assistant_content(msg.content)})
            else:
                result.append({"role": msg.role, "content": msg.content})
        return result

    def to_tools(self, tools: dict[str, Tool]) -> list[dict[str, Any]]:
        return [
            {
                "name": tool.name,
                "description": tool.description,
                "input_schema": self._input_schema_for(tool.parameters),
            }
            for tool in tools.values()
        ]

    @staticmethod
    def _input_schema_for(parameters: dict[str, Any]) -> dict[str, Any]:
        properties: dict[str, Any] = {}
        required: list[str] = []

        for name, spec in parameters.items():
            prop = {"type": spec.get("type"), "description": spec.get("description")}
            properties[name] = {k: v for k, v in prop.items() if v is not None}
            if spec.get("required", True):
                required.append(name)

        return {"type": "object", "properties": properties, "required": required}

    def to_payload(
        self, context: Context, *, max_output_tokens: int = 1024, tools: list[dict[str, Any]] | None = None
    ) -> dict[str, Any]:
        return {
            "model": self.model,
            "system": context.system,
            "max_tokens": max_output_tokens,
            "tools": self.to_tools(context.tools) if tools is None else tools,
            "messages": self.to_messages(context.messages),
        }

    @property
    def headers(self) -> dict[str, str]:
        return {
            "Content-Type": "application/json",
            "x-api-key": self.api_key,
            "anthropic-version": "2023-06-01",
        }

    @property
    def url(self) -> str:
        return self.BASE_URL

    # Normalizes an Anthropic Messages API response into the common shape
    # (see Backends::Base for the full content-block contract). Anthropic's
    # native thinking/redacted_thinking blocks are mapped to "reasoning"
    # blocks, preserving the signature so they can be echoed back unchanged
    # (the API rejects modified thinking blocks when continuing on the same
    # model).
    def parse_response(self, response: dict[str, Any]) -> dict[str, Any]:
        stop_reason = "tool_use" if response.get("stop_reason") == "tool_use" else "end_turn"
        content = [self._normalize_block(block) for block in response.get("content") or []]
        return {"stop_reason": stop_reason, "content": content}

    @staticmethod
    def _normalize_block(block: dict[str, Any]) -> dict[str, Any]:
        if block.get("type") == "thinking":
            return {"type": "reasoning", "text": str(block.get("thinking") or ""), "signature": block.get("signature")}
        if block.get("type") == "redacted_thinking":
            return {"type": "reasoning", "text": "", "redacted": True, "signature": block.get("data")}
        return block

    # Rebuilds Anthropic assistant content from normalized blocks (the inverse
    # of parse_response). Text-only turns are stored as a bare String and pass
    # through unchanged; "reasoning" blocks are re-emitted as native
    # thinking/redacted_thinking blocks so signatures round-trip intact.
    @classmethod
    def _assistant_content(cls, content: str | list[dict[str, Any]]) -> str | list[dict[str, Any]]:
        if isinstance(content, str):
            return content
        return [cls._denormalize_block(block) for block in content]

    @staticmethod
    def _denormalize_block(block: dict[str, Any]) -> dict[str, Any]:
        if block.get("type") != "reasoning":
            return block

        if block.get("redacted"):
            return {"type": "redacted_thinking", "data": block.get("signature")}
        return {"type": "thinking", "thinking": str(block.get("text") or ""), "signature": block.get("signature")}
